include/deal.II/grid/grid_tools.h

```diff
@@ -2089,6 +2102,128 @@ namespace GridTools
+  Point<Iterator::AccessorType::space_dimension>
+  project_to_object(const Iterator &object,
+                    const Point<Iterator::AccessorType::space_dimension> &trial_point)
+  {
+    const unsigned int spacedim = Iterator::AccessorType::space_dimension;
+    const unsigned int structdim = Iterator::AccessorType::structure_dimension;
+
+    if (structdim == 1 && spacedim <= 1)
+      return trial_point;
+    else if (structdim == 2 && spacedim <= 2)
+      return trial_point;
+    else if (structdim == 3 && spacedim <= 3)
+      return trial_point;
+    else if (structdim == 1)
+      {
+        // find the point that lies on
+        // the line p1--p2. the
+        // formulas pan out to
+        // something rather simple
+        // because the mapping to the
+        // line is linear
+        const Point<spacedim> p1 = object->vertex(0),
+                              p2 = object->vertex(1);
+        const double s = (trial_point-p1)*(p2-p1) / ((p2-p1)*(p2-p1));
+        return p1 + s*(p2-p1);
+      }
+    else if (structdim == 2)
+      {
+        // all points on the surface are given by
+        //   x(\xi) = sum_i v_i phi_x(\xi)
+        // where v_i are the vertices of the quad,
+        // and \xi=(\xi_1,\xi_2) are the reference
+        // coordinates of the quad. so what we are
+        // trying to do is find a point x on
+        // the surface that is closest to the point
+        // y. there are different ways
+        // to solve this problem, but in the end
+        // it's a nonlinear problem and we have to
+        // find reference coordinates \xi so that
+        //   J(\xi) = 1/2 || x(\xi)-y ||^2
+        // is minimal. x(\xi) is a function that
+        // is structdim-linear in \xi, so J(\xi) is
+        // a polynomial of degree 2*structdim that
+        // we'd like to minimize. unless structdim==1,
+        // we'll have to use a Newton
+        // method to find the
+        // answer. This leads to the
+        // following formulation of
+        // Newton steps:
+        //
+        // Given \xi_k, find \delta\xi_k so that
+        //   H_k \delta\xi_k = - F_k
+        // where H_k is an approximation to the
+        // second derivatives of J at \xi_k, and
+        // F_k is the first derivative of J.
+        // We'll iterate this a number of times
+        // until the right hand side is small
+        // enough. As a stopping criterion, we
+        // terminate if ||\delta\xi||<eps.
+        //
+        // As for the Hessian, the best choice
+        // would be
+        //   H_k = J''(\xi_k)
+        // but we'll opt for the simpler
+        // Gauss-Newton form
+        //   H_k = A^T A
+        // i.e.
+        //   (H_k)_{nm} = \sum_{i,j} v_i*v_j *
+        //                   \partial_n phi_i *
+        //                   \partial_m phi_j
+        // we start at xi=(0.5,0.5).
+        Point<structdim> xi;
+        for (unsigned int d=0; d<structdim; ++d)
+          xi[d] = 0.5;
```

Comment:
that's probably good enough, but we have a function somewhere that approximates a quad as a plane and finds the closest point on that plane. that is probably pretty good in general, and may give a better starting point.